### PR TITLE
Add Client method

### PR DESCRIPTION
**Pull request by gmlewis**
Fixes: #1859.
Closes: #2015.

### CODE REVIEW

**Comment by gmlewis** [comment]
@bluekeyes and @jeffmendoza - what do you think of this solution?  Feel free to LGTM+Approve if you agree.

**Review by jeffmendoza — APPROVED** [review]

File: github/github.go [hunk 1]
```diff
@@ -188,6 +188,13 @@ type service struct {
 	client *Client
 }
 
+// Client returns the http.Client used by this GitHub client.
+func (c *Client) Client() *http.Client {
+	c.clientMu.Lock()
+	defer c.clientMu.Unlock()
+	return c.client
```
[the remark this answers is not in the document]
Comment (gmlewis) [reply github/github.go:195]: No, I don't think that is necessary.  The point of the lock is to prevent two clients from attempting to overwrite the same variable at the same time (for handling redirects).  A shallow copy is not necessary.

The point of solving #1859 is to use the same `http.Client` for older versions of this repo and it can be used concurrently without the need for a shallow copy.

[the remark this answers is not in the document]
Comment (bluekeyes) [reply github/github.go:195]: Hmm, I was thinking of a case like this (admittedly contrived):

```go
import (
    "context"
    "github.com/google/go-github/v37/github"
    github36 "github.com/google/go-github/v36/github"
)

func useTwoVersionsConcurrently(ctx context.Context, client *github.Client) {
    client36 := github36.NewClient(client.Client())
    client36.BaseURL = client.BaseURL

    go func() {
        client.DownloadReleaseAsset(ctx, "octocat", "Hello-World", 1, http.DefaultClient)
    }()

    go func() {
        client36.MigrationArchiveURL(ctx, "ORG", 42)
    }()
}
```

I thought `client.DownloadReleaseAsset` and `client36.MigrationArchiveURL` would acquire different locks and then modify the `CheckRedirect` function on the same `http.Client` instance (the two GitHub clients share a pointer.) But I could easily be overlooking or misunderstanding something.

My understanding is that the `Transport` field is the most important piece to share between GitHub clients, as this has the authentication, caching, and other persistent state. The `http.Client` doesn't appear to have much state of its own.

[the remark this answers is not in the document]
Comment (gmlewis) [reply github/github.go:195]: @bluekeyes - I understand what you are saying, but completely disagree.  What you wrote should work fine.

According to: https://pkg.go.dev/net/http

> Clients and Transports are safe for concurrent use by multiple goroutines and for efficiency should only be created once and re-used.

Therefore, it is safe to use a single `http.Client` concurrently.  _That_ is not the purpose for the mutex and is not the problem being solved by it.

The purpose of the mutex is to prevent a "race condition" and is there to protect the actual variable `Client.client` itself.  According to [Wikipedia](https://en.wikipedia.org/wiki/Race_condition),

> A race condition arises in software when a computer program, to operate properly, depends on the sequence or timing of the program's processes or threads. Critical race conditions cause invalid execution and software bugs. Critical race conditions often happen when the processes or threads depend on some shared state. Operations upon shared states are done in critical sections that must be mutually exclusive. Failure to obey this rule can corrupt the shared state.

Some of our methods temporarily swap out the value of `Client.client` in order to perform some cool features like following redirects to get the final values so that the users of this client library don't have to figure all that stuff out for themselves.

If the mutex were not there, it would be certainly possible for two concurrent goroutines to both attempt to swap out a new temporary client by both messing with the `Client.client` variable directly, and due to the race condition, you would have no way of knowing what value would be used or how it would end up.  The mutex prevents this race condition.

Therefore, I stand by my assertion that this PR should be a valid solution to #1859.

[the remark this answers is not in the document]
Comment (bluekeyes) [reply github/github.go:195]: Thanks for taking the time to discuss this with me, @gmlewis. I hope I'm not being too difficult - I use this library a lot and really appreciate the time and effort you put into maintaining it, so I only want to help fix what I believe is a bug.

To that end, Go's race detector also indicates that the [solution as currently implemented has a data race](https://gist.github.com/bluekeyes/e50fd4c695b7079e47b0821c246b96d1#gistcomment-3823015).

> The purpose of the mutex is to prevent a "race condition" and is there to protect the actual variable `Client.client` itself.
> [...]
> Some of our methods temporarily swap out the value of `Client.client` in order to perform some cool features [...]

I think this is the source of our disagreement. Looking through the library, the two functions ([1](https://github.com/google/go-github/blob/master/github/repos_releases.go#L288-L297), [2](https://github.com/google/go-github/blob/master/github/migrations.go#L170-L180)) that hold `client.clientMu` update the `Client.client.CheckRedirect` field, not the `Client.client` field. If parts of the library do change the `Client.client` field, they seem to do so without holding the lock.

In other words, the current code uses `clientMu` to prevent concurrent modification of the `CheckRedirect` field. By returning a pointer to the `Client.client`, external code can now modify the same `CheckRedirect` field without holding the lock (which is private to the original `Client`.)

[the remark this answers is not in the document]
Comment (gmlewis) [reply github/github.go:195]: Wow, awesome, @bluekeyes !!!  Looks like I'm wrong!  I'm in a meeting... I'll take a look in a bit.  Thank you!

[the remark this answers is not in the document]
Comment (gmlewis) [reply github/github.go:195]: I totally stand corrected... thank you for your persistence, @bluekeyes !!!  I really appreciate it.

Have you tried running the race test with your shallow copy version?  If it passes the race test, then I should be able to update this PR if you agree.

[the remark this answers is not in the document]
Comment (bluekeyes) [reply github/github.go:195]: Glad we figured things out! Using a library version that makes a shallow copy of the client passes the race detector test on my machine, so I think that will work fine (unless you or anyone else spots another issue.)

**Comment by gmlewis** [comment]
@bluekeyes - feel free to LGTM+Approve this and we can get it merged.

**Review by bluekeyes — APPROVED** [review]
LGTM!
